### ISSUE TITLE
Change the way generated clients use HttpClient - Potential socket exhaustion

### ISSUE THREAD

**Issue by ishepherd** [RSuter/NSwag#1097]
According to https://aspnetmonsters.com/2016/08/2016-08-27-httpclientwrong/

After each call, even though you've disposed the HttpClient, Windows holds the socket(s) it was using open for 240 seconds. There are only thousands of sockets available, so you can achieve socket exhaustion quite easily.

> Although [HttpClient] implements the IDisposable interface it is actually a shared object. This means that under the covers it is reentrant) and thread safe. Instead of creating a new instance of HttpClient for each execution you should share a single instance of HttpClient for the entire lifetime of the application.
> ...
> 1. Make your HttpClient static.
> 2. Do not dispose of or wrap your HttpClient in a using unless you explicitly are looking for a particular behaviour (such as causing your services to fail).

**Comment by ishepherd** [comment]
Reading further...

Although [some MSDN agrees...](https://docs.microsoft.com/en-us/aspnet/web-api/overview/advanced/calling-a-web-api-from-a-net-client#create-and-initialize-httpclient)
> HttpClient is intended to be instantiated once and reused throughout the life of an application. The following conditions can result in SocketException errors:
> * Creating a new HttpClient instance per request.
> * Server under heavy load.
> 
> Creating a new HttpClient instance per request can exhaust the available sockets.

...it can create another: You don't want to keep a socket open forever, for example if there is a DNS flip to a different deployment, you will still be talking to the old one.

Some people have custom logic so that both issues are solved. http://disq.us/p/1d4okj9 https://github.com/dotnet/corefx/issues/11224#issuecomment-347361456

**Comment by ishepherd** [comment]
OK - Sorry, I realize you're way ahead of me on this. :-)
I see checkboxes in NSwagStudio that allow control of HttpClient life cycle.

Please close this issue if not required. Thanks.

**Comment by RSuter** [comment]
My problem with HttpClient is that there is no single or commonly used best practice... This is why I it is currently up to the user to correctly use it :-). The question: What is the best approach to solve this problem?

As you may know there are currently three ways to handle HttpClient in the generated code: 

- Create an HttpClient per request (no injection, default)
- Inject HttpClient, handled by client
- Change the HttpClient type, to e.g. "MyNamespace.IHttpClient" (the interface just needs the required methods which are generated). 

Maybe we should implement a separate library which provides which provides HttpClient proxy implementations which internally manage HttpClient based on time or whatever - the library could provide different implementation which can be chosen depending on the use case/scenario?

**Comment by ishepherd** [comment]
I'll bet the commonly used 'best' practice is

> * ~Create an HttpClient per request (no injection, default)~

Devs think "I Disposed it, I did the right thing!", not realizing that creating too many HttpClients within 240 seconds will break their system...

My conclusion is that HttpClients should *always be static* (per my [MSDN link](https://docs.microsoft.com/en-us/aspnet/web-api/overview/advanced/calling-a-web-api-from-a-net-client#create-and-initialize-httpclient)) but, because this might lead to it reusing a single socket forever, *kept in a pool with round-robin and expiry time*.

The pool would need a couple of parameters
  * Size (Reasonable default - 8?)
  * Expiry (Reasonable default - 10 seconds?)

I think this would work for everyone?

Obviously I will be looking for a prebuilt pool class... will have a dig around

Apparently this is faster as well as safer: https://github.com/mspnp/performance-optimization/blob/master/ImproperInstantiation/docs/LoadTesting.md

**Comment by RSuter** [comment]
So is there a widely used NuGet package with a good HttpClient wrapper which we could use or propose to use?

**Comment by ishepherd** [comment]
@RSuter I haven't had time to look into this any further. Hopefully next month.

**Comment by bulfonjf** [comment]
How about this wrapper? https://github.com/NimaAra/Easy.Common/blob/master/Easy.Common/RestClient.cs

**Comment by ishepherd** [comment]
You could code a basic version directly in the generator to avoid sprouting another dependency. What we're now doing is

```
private static readonly ObjectWithLifespan<HttpClient> OurHttpClient = new ObjectWithLifespan<HttpClient>(
    lifespan: TimeSpan.FromSeconds(30),
    creator: () => new HttpClient());

public class ObjectWithLifespan<T>
{
    private class Box
    {
        public T Instance;
        public DateTime Expiry;
    }

    private Box Instance;
    private readonly TimeSpan Lifespan;
    private readonly Func<T> Creator;

    /// <summary>
    /// Function provided must be safe. It might be invoked multiple times needlessly or concurrently,
    /// the value you get may not be from the latest invocation
    /// </summary>
    public ObjectWithLifespan(TimeSpan lifespan, Func<T> creator)
    {
        Lifespan = lifespan;
        Creator = creator ?? throw new ArgumentNullException(nameof(creator));
    }

    public T Value
    {
        get
        {
            var current = Instance;
            if (current == null || current.Expiry <= DateTime.UtcNow)
            {
                current = new Box { Instance = Creator(), Expiry = DateTime.UtcNow + Lifespan };
                Instance = current;
            }
            return current.Instance;
        }
    }

}
```

**Comment by RSuter** [comment]
I think we should provide an implementation for IHttpClient in a NuGet package so that you can specify IHttpClient instead of HttpClient in the NSwag settings and use this NuGet package with the special implementation...

**Comment by ishepherd** [comment]
Whatever approach, I think it should be the default. I don't see that all the confusion around this is really justified, it oughta be completely possible to have a 'right' solution

> @bulfonjf How about this wrapper? https://github.com/NimaAra/Easy.Common/blob/master/Easy.Common/RestClient.cs

depends on ServicePointManager - looks like it's problematic on .NET Core... https://github.com/dotnet/corefx/issues/11224

**Comment by Eneuman** [comment]
Now that we have `services.AddHttpClient` in .Net Core 2.1+, I would suggest changing the C# client template in the following way:

The public constructor:
```
        public ProcessmotorClient(System.Net.Http.HttpClient httpClient)
        {
            BaseUrl = baseUrl; 
            _httpClient = httpClient; 
            _settings = new System.Lazy<Newtonsoft.Json.JsonSerializerSettings>(() => 
            {
                var settings = new Newtonsoft.Json.JsonSerializerSettings();
                UpdateJsonSerializerSettings(settings);
                return settings;
            });
        }
```

is changed to 
```
        public ProcessmotorClient(string baseUrl, System.Net.Http.HttpClient httpClient)
        {
            _httpClient = httpClient; 
            _settings = new System.Lazy<Newtonsoft.Json.JsonSerializerSettings>(() => 
            {
                var settings = new Newtonsoft.Json.JsonSerializerSettings();
                UpdateJsonSerializerSettings(settings);
                return settings;
            });
        }
```
BaseUrl is removed in favor for `httpClient.BaseAddress`

This change will make it possible to do this without adding any extra code:

```
    services.AddHttpClient<MyClient>(client =>
    {
        client.BaseAddress = new Uri("https://api.myclient.com/");
    });
```

Using AddHttpClient this way also solves problems with port starvation and DNS lookup issues in Micro Services.

**Comment by RSuter** [comment]
Have you tried to disable the UseBaseUrl setting?

![image](https://user-images.githubusercontent.com/2603405/52772818-c05b7880-3039-11e9-86ce-8c8201a6167c.png)

Produces: 

![image](https://user-images.githubusercontent.com/2603405/52772857-db2ded00-3039-11e9-94fa-4e04521549a3.png)

**Comment by Eneuman** [comment]
Nice :)
I didn't know this already existed. 
Can this also be set as a parameter in .projcs when using the preview of NSwag.MSBuild.CodeGeneration?
(I'm trying to automaticly create a Nuget Package of the client each time I check in code to Azure evOps)

**Comment by RSuter** [comment]
I think with the NSwagOptions attribute you can pass /UseBaseUrl:false - but not sure if the attribute is called like this...

**Comment by Eneuman** [comment]
It worked great! Thanks!
I added the following parameter:  GenerateNSwagCSharpOptions="/UseBaseUrl:true"

**Comment by RSuter** [comment]
BTW: NSwag.MSBuild.CodeGeneration has been unlisted/renamed to NSwag.ApiDescription.Design

**Comment by stephengtuggy** [comment]
In .NET Core 2.1 and up, I believe the preferred pattern is to use `IHttpClientFactory.CreateClient()` every time you need an HttpClient instance. The default factory implementation then creates HttpClient instances that clean up their underlying sockets and other resources after 2 minutes.

What about turning `_httpClient` into a read-only, private property? Then in the `property get`, on .NET Core 2.1+ you could call IHttpClientFactory.CreateClient(), while on .NET Framework you could just use a static HttpClient instance, or else a custom algorithm that imitates the IHttpClientFactory behaviour. Though the initialization of generated clients might be complicated if you supported all of the above.

Thoughts?

**Comment by SebastianStehle** [comment]
I think it would make more sense to have a custom IHttpClientProvider interface. Then we can implement this in a backwards compatible way with a `StaticHttpClientProvider" implementation that is used under the hood.

**Comment by stephengtuggy** [comment]
> I think it would make more sense to have a custom IHttpClientProvider interface. Then we can implement this in a backwards compatible way with a `StaticHttpClientProvider" implementation that is used under the hood.

Yes. This is what I am now leaning towards. I have implemented this myself in a project or two.

**Comment by stephengtuggy** [comment]
And maybe give your IHttpClientProvider interface a read-only, boolean property that indicates whether or not user code should Dispose() of the IHttpClient it got after it's done with it. `ShouldDisposeAfterUse` or something like that.

**Comment by SebastianStehle** [comment]
Or a release method. It is a typical pattern for object pooling. See ObjectPool class
